st2actions/tests/unit/test_pythonrunner.py

```diff
@@ -111,6 +112,52 @@ def test_action_with_user_supplied_env_vars(self, mock_popen):
             else:
                 self.assertEqual(actual_env[key], value)
 
+    @mock.patch('st2actions.runners.pythonrunner.subprocess.Popen')
+    def test_stdout_interception_and_parsing(self, mock_popen):
+        values = {'delimiter': ACTION_OUTPUT_RESULT_DELIMITER}
+
+        # No output to stdout and no result (implicit None)
+        mock_stdout = '%(delimiter)sNone%(delimiter)s' % values
+        mock_stderr = 'foo stderr'
+        mock_process = mock.Mock()
+        mock_process.communicate.return_value = (mock_stdout, mock_stderr)
+        mock_process.returncode = 0
+        mock_popen.return_value = mock_process
+
+        runner = pythonrunner.get_runner()
+        runner.action = self._get_mock_action_obj()
+        runner.runner_parameters = {}
+        runner.entry_point = PACAL_ROW_ACTION_PATH
+        runner.container_service = service.RunnerContainerService()
+        runner.pre_run()
+        (_, output, _) = runner.run({'row_index': 4})
+
+        self.assertEqual(output['stdout'], '')
+        self.assertEqual(output['stderr'], mock_stderr)
+        self.assertEqual(output['result'], 'None')
+        self.assertEqual(output['exit_code'], 0)
+
+        # Output to stdout and no result (implicit None)
+        mock_stdout = 'pre result%(delimiter)sNone%(delimiter)spost result' % values
+        mock_stderr = 'foo stderr'
+        mock_process = mock.Mock()
+        mock_process.communicate.return_value = (mock_stdout, mock_stderr)
+        mock_process.returncode = 0
+        mock_popen.return_value = mock_process
+
+        runner = pythonrunner.get_runner()
+        runner.action = self._get_mock_action_obj()
+        runner.runner_parameters = {}
+        runner.entry_point = PACAL_ROW_ACTION_PATH
+        runner.container_service = service.RunnerContainerService()
+        runner.pre_run()
+        (_, output, _) = runner.run({'row_index': 4})
+
+        self.assertEqual(output['stdout'], 'pre resultpost result')
+        self.assertEqual(output['stderr'], mock_stderr)
+        self.assertEqual(output['result'], 'None')
```

Comment:
As far as `ast.safe_eval` goes - I'm not too sure about it, since we should enforce good explicit practices and require users to return a JSON which is a common currency for structured results in actions.

Using `ast.safe_eval` is just supporting laziness and more implicit and magical code (although, we already do this in multiple places).